src/Razor/src/Microsoft.VisualStudio.LanguageServices.Razor/LanguageClient/Cohost/CohostDocumentCompletionEndpoint.cs

```diff
@@ -0,0 +1,295 @@
+                            solutionInfo,
+                            razorDocument.Id,
+                            completionContext,
+                            ToVsLSP<Position>(request.Position).AssumeNotNull(),
+                            cancellationToken),
+                cancellationToken).ConfigureAwait(false) is not { } completionPositionInfo)
+        {
+            // If we can't figure out position info for request position we can't return completions
+            return null;
+        }
+
+        var documentPositionInfo = completionPositionInfo.DocumentPositionInfo;
+        if (documentPositionInfo.LanguageKind != RazorLanguageKind.Razor)
+        {
+            completionContext = DelegatedCompletionHelper.RewriteContext(completionContext, documentPositionInfo.LanguageKind);
+        }
+
+        // First of all, see if we in HTML and get HTML completions before calling OOP to get Razor completions.
+        // Razor completion provider needs a set of existing HTML item labels.
+
+        VSInternalCompletionList? htmlCompletionList = null;
+        var razorCompletionOptions = new RazorCompletionOptions(
+            SnippetsSupported: true, // always true in non-legacy Razor
+            AutoInsertAttributeQuotes: clientSettings.AdvancedSettings.AutoInsertAttributeQuotes,
+            CommitElementsWithSpace: clientSettings.AdvancedSettings.CommitElementsWithSpace);
+        using var _ = HashSetPool<string>.GetPooledObject(out var existingHtmlCompletions);
+
+        if (CompletionTriggerCharacters.IsValidTrigger(CompletionTriggerCharacters.HtmlTriggerCharacters, completionContext))
+        {
+            // We can just blindly call HTML LSP because if we are in C#, generated HTML seen by HTML LSP may return
+            // results we don't want to show. So we want to call HTML LSP only if we know we are in HTML content.
+            if (documentPositionInfo.LanguageKind == RazorLanguageKind.Html)
+            {
+                htmlCompletionList = await GetHtmlCompletionListAsync(request, razorDocument, razorCompletionOptions, cancellationToken);
+                if (htmlCompletionList is not null)
+                {
+                    existingHtmlCompletions.UnionWith(htmlCompletionList.Items.Select(i => i.Label));
+                }
+            }
+        }
+
+        _logger.LogDebug($"Calling OOP to get completion items at {request.Position} invoked by typing '{request.Context?.TriggerCharacter}'");
+
+        var data = await _remoteServiceInvoker.TryInvokeAsync<IRemoteCompletionService, Response>(
+            razorDocument.Project.Solution,
+            (service, solutionInfo, cancellationToken)
+                => service.GetCompletionAsync(
+                        solutionInfo,
+                        razorDocument.Id,
+                        completionPositionInfo,
+                        completionContext,
+                        _clientCapabilities.AssumeNotNull(),
```

Comment:
I agree, I believe we can reduce the size.

Reply:
Double-checked, we do need that one boolean (supports VS extensions), plus clientCapabilities.TextDocument?.Completion.

Reply:
Since we need more than just bool, and also that bool (supportsVSExtensions) is now needed in other Roslyn APIs (e.g. document symbols), seem that passing in as client initialization options is the right thing to do. I created a subtask [https://github.com/dotnet/razor/issues/11102](https://github.com/dotnet/razor/issues/11102) and will address this perf improvement separately.